src/storage/txn/commands/mod.rs

```diff
@@ -432,3 +469,23 @@ impl Debug for Command {
         self.command_ext().fmt(f)
     }
 }
+
+pub trait ReadCommand<E: Engine>: CommandExt {
+    fn process_read(
+        &mut self,
+        snapshot: E::Snap,
+        statistics: &mut Statistics,
+    ) -> Result<ProcessResult>;
+}
+
+pub trait WriteCommand<S: Snapshot, L: LockManager, P: PdClient + 'static>: CommandExt {
+    fn process_write(
+        &mut self,
+        snapshot: S,
+        lock_mgr: &L,
+        pd_client: Arc<P>,
+        extra_op: ExtraOp,
+        statistics: &mut Statistics,
+        pipelined_pessimistic_lock: bool,
```

Comment:
I think we can probably factor these args out into a struct (or maybe two structs - I feel like the snapshot, lock manager, and pd client, are some abstraction over the storage layer, and the rest are some kind of context)